File: 0724. Find Pivot Index/0724.java
// time complexity: O(n)
// space complexity: O(1)
class Solution {
    // single‐pass with running sums
    public int pivotIndex(int[] nums) {
        // compute total sum of array
        int total = 0;
        for (int num : nums) total += num;
        int leftSum = 0;
        // scan once: at each i, rightSum = total − leftSum − nums[i]
        for (int i = 0; i < nums.length; i++) {
            if (leftSum == total - (leftSum + nums[i])) {
                return i;  // found the pivot
            }
            leftSum += nums[i];  // include nums[i] into leftSum for next iteration
        }
        return -1;  // no pivot found
    }
}

File: 0724. Find Pivot Index/0724.py
# time complexity: O(n)
# space complexity: O(1)
class Solution(object):
    def pivotIndex(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        total = sum(nums)     # total sum
        left_sum = 0
        for i, x in enumerate(nums):
            # right_sum = total − (left_sum + current)
            if left_sum == total - (left_sum + x):
                return i
            left_sum += x
        return -1